### PR TITLE
The Great Waterproofing

### PR DESCRIPTION
**Pull request by Ilysen**
#### Summary
 ```SUMMARY: Balance "Rainproofing and waterproofing has been rebalanced across the board for almost all clothing."```

#### Purpose of change
A great deal of armor lacks waterproofing when it shouldn't, like things made out of thick leather or steel, and some things provide complete rainproofing even though they only cover one or two body parts.

#### Describe the solution
The `RAINPROOF` flag has been reworked. Instead of protecting the entire body from rain, it now protects only the covered body parts. Additionally, two new flags have been added:
* `RAINBLOCK` prevents rain from getting through a piece of clothing on 70% of turns.
* `RAINRESIST` prevents rain from getting through on 30% of turns.

The `WATERPROOF` flag remains able to block all rain 100% of the time. Because of these flag changes, a large amount of changes have been made to existing armors, with some flags removed and others added. Overall, there are no more magical anti-rain cargos, but much more gear should be at least partially rainproof.

#### Describe alternatives you've considered

Giving materials innate water resistance. Because something's material doesn't determine its actual construction (like a windbreaker being made of plastic is different from knee pads made of plastic), I decided not to do this - at least until further steps are taken.

#### Changed clothing

I followed a few general guidelines - metal armor, plastic/Kevlar gear, and most leather is rainproof. The aforementioned gear but with low coverage (metal arm guards, skid lid, etc) or very specialized cotton (i.e. beekeeping suit) has been made rain-blocking. Thick cotton, like hoodies, is rain-resistant.

Any clothing that wouldn't actively drench your body while you're wearing it, or that feels fine when wet, is water-friendly.

After this PR, flags exist on the following items. Duplicates do exist, and keep in mind that anything that's waterproof is also rainproof.

There are more than this, but I ran out of energy to continue listing them at line 9000. Sue me.

* ***Rainproof:*** ten-gallon hat, steel arm guards, boiled leather armor, bone body armor, leather body armor, plate armor, lorica segmentata, plated leather armor, wool beret, all chainmail armor except coif, fur cloak, rain coat (which now covers legs), gutskin parka, cowboy hat, cuirass, fur duster, faux fur duster, leather duster, survivor duster, fencing mask, bone armor gauntlets, chitinous gauntlets, leather armor gauntlets, bag gloves, boonie hat, sombrero, barbute helmet, leather armor helmet, emergency jacket, leather jacket, armored leather jacket, red leather jacket, windbreaker, steel leg guards, porkpie hat, pot helmet, shark suit, faraday shark suit, sleeping bag, fur sleeping bag, sleeveless fur duster, sleeveless faux fur duster, sleeveless leather duster, sleeveless fur trenchcoat, sleeveless faux fur trenchcoat, sleeveless leather trenchcoat, straw hat
* ***Rain-blocking:*** metal arm guards, lamellar cuirass, nomad gear, beekeeping suit, chainmail coif, cleansuit, cloak, wool cloak, fur coat, faux fur coat, sable coat, greatcoat, baseball cap, fur hat, faux fur hat, army jacket, flannel jacket, jean jacket, iron greaves, army pants, fur pants, leather pants, peacoat, 
* ***Rain-resistant:*** light nomad gear, scrap suit, giant novelty backpack, balaclava, beekeeping gloves, beekeeping hood, beekeeping suit, blanket, bondage mask and suit, bowler hat, cassock, dinosaur suit, fencing jacket, fencing pants, gambeson, skid lid, hoodie, housecoat, light jacket, jeans, red jeans, kittel, cargo pants, wool poncho,
* ***Water-friendly:*** 2-by-arm guards, 2-by-shin guards, leather apron, arm splint, leg splint, arm splint XL, leg splint XL, bone arm guards, chitin arm guards, hard arm guards, leather arm guards, neoprene arm sleeves, boiled leather armor, bone body armor, leather body armor, lorica segmentata, camo tank top, basketball shorts, leather backpack, all badges, all sheaths/scabbards/holsters, beret, all bikinis, bindle, blindfold, all chainmail armor, leather chaps, leather corset, golden crown, dive bag, diver's watch, duffel bag, tactical dump pouch, elbow pads, knee pads, all glasses, fanny pack, all belts, fingerless gloves, armored fingerless gloves, survivor wetsuit gloves, light gloves, glove liners, medical gloves, tactical gloves, winter gloves, go bag, swim goggles, survivor wetsuit, halter top, helmet liner, hide bags, survivor wetsuit hood, all hot pants, snow goggles, jerrypack, all judo belts, leather pouch, leggings, hard leg guards, drop leg pouches, makeshift knapsack, makeshift sling, ballistic mask, Guy Fawkes mask, hockey mask, messenger bag, mittens, MOLLE pack, mouthpiece, all flame-resistant gear, plastic shopping bag, purse, all quivers, pouch, military rucksack, runner pack, fancy watch, shark suit, faraday shark suit, sling pack, straw basket,
* ***Completely waterproof:*** AEP suit, chitinous armor, fur body armor, O-yoroi, scavenger gear, boxing gloves, briefcase, all turnout gear, leather cloak, lab coat, winter coat, all power armor, dragon skin vest, fire gauntlets, fire helmet, fishing waders, floatation vests, football armor, survivor firesuit, survivor firegloves, heavy survivor gloves, leather gloves, light survivor gloves, armored gauntlets, rubber gloves, survivor gloves, work gloves, winter survivor gloves, XL survivor gloves, ski goggles, hard hat, hooded hard hat, headgear, army helmet, baseball helmet, barbute helmet, bike helmet, bone armor helmet, chitinous helmet, conical helm, football helmet, galea, heavy survivor helmet, kabuto, lobster helmet, motorcycle helmet, nasal helm, nomad cowl, great helm (which is also no longer wearable with other helmet, what the heck), scavenger cowl, survivor helmet, XL survivor helmet, light survivor hood, rain hood, survivor hood, winter survivor hood, XL survivor hood, heavy survivor suit, Kevlar vest, light survivor body armor, light survivor cargo pants, light survivor suit, motorcycle armor, motorcycle pants, FB51 optical cloak, ski pants, survivor cargo pants, pickelhaube, sleeveless survivor trenchcoat, WS-47G moisture retention suit,

### CODE REVIEW

File: data/json/items/armor.json [hunk 1]
```diff
@@ -992,7 +992,7 @@
     "warmth": 25,
     "material_thickness": 2,
     "environmental_protection": 1,
-    "flags": [ "VARSIZE", "OUTER" ]
+    "flags": [ "VARSIZE", "OUTER", "RAINPROOF", "WATERPROOF" ]
```
Comment (Night-Pryanik) [inline data/json/items/armor.json]: Are you sure this cotton suit is rainproof?

File: data/json/items/armor.json [hunk 2]
```diff
@@ -1693,7 +1693,7 @@
     "coverage": 95,
     "encumbrance": 20,
     "material_thickness": 4,
-    "flags": [ "VARSIZE", "STURDY", "HELMET_COMPAT" ]
+    "flags": [ "VARSIZE", "STURDY", "HELMET_COMPAT", "RAINPROOF", "WATER_FRIENDLY" ]
```
Comment (Night-Pryanik) [inline data/json/items/armor.json]: Chainmail suit, full of holes, is rainproof?

File: data/json/items/armor.json (hunk 1 above)
Comment (Ilysen) [reply data/json/items/armor.json]: Beekeeping suits are very well-sealed and have full coverage, plus the cotton is rather thick. I thought it was okay to rainproof them, but I can remove that if you feel it's too unrealistic.

File: data/json/items/armor.json (hunk 2 above)
Comment (Ilysen) [reply data/json/items/armor.json]: I actually figured the rainproofing to be the leather beneath the suit, but the metal itself is also very tightly linked.

**Comment by Night-Pryanik** [comment]
There is a major misunderstanding in mechanics of `RAINPROOF` flag. Quoting @Coolthulhu from https://github.com/CleverRaven/Cataclysm-DDA/issues/19732#issuecomment-289432343:
> RAINPROOF prevents 98% of applications of rain, regardless of body part (ie. if your socks are RAINPROOF, you are rainproof).

https://github.com/CleverRaven/Cataclysm-DDA/blob/d02530346d379624865f855e7263d451ac894ad1/src/weather.cpp#L371-L378

So only full-body suits like ANBC should have `RAINPROOF` flag. Power armor should be rainproof only with a helmet.

The description of this flag in `JSON_FLAGS.md` is erroneously stating that this flag will protect covered body parts instead of a whole body. It must be fixed.

**Comment by Ilysen** [comment]
A rework of the `RAINPROOF` flag was going to be part 2 of this initiative, but I can remove the flags from non-full-body things for now.

File: data/json/items/armor.json [hunk 3]
```diff
@@ -7302,6 +7305,7 @@
     "encumbrance": 50,
     "warmth": 5,
     "material_thickness": 2,
+    "flags": [ "WATERPROOF" ]
```
Comment (pjf) [inline data/json/items/armor.json]: Looks like we're missing a comma here!

```suggestion
    "flags": [ "WATERPROOF" ],
```

**Comment by I-am-Erk** [comment]
This should definitely have a list of changes in the PR description.

**Comment by Photoloss** [comment]
>     * Gear that water wouldn't get through or that water would be repelled from has been made `WATERPROOF`. This includes very thick cotton things like army jackets.

What exactly is the criterion here? Most cloth items will simply absorb a certain amount of water letting virtually none through until saturated, even a regular sweater/hoodie can handle some light rainfall. Meanwhile something like rubber boots, medical gloves or a full hazmat suit is literally waterproof and will not let *any* amount of water through as long as you seal the edge of the garment. Metal armour is somewhere in between where the plates themselves are 100% waterproof but the gaps provide virtually no protection if a splash comes in at the correct angle.

**Comment by Ilysen** [comment]
Further differentiation is something I'm aiming to implement in part 2 (so as to avoid having gigantic PRs), but for now, very thick cotton like army clothes are waterproof, thick leather is waterproof, and metal armor is waterproof. Things like the cuirass can technically not absorb water in the joints, but simulating that at this state seems a bit silly - at least until I can start on the next stage.

**Comment by Ilysen** [comment]
After skimming through, it's apparent that removing the `RAINPROOF` flag from things right now would more or less remove all rainproof clothing from the game. Is it okay if I remove them from the ones changed in this PR, but mark those armor pieces with comments with a todo?

**Comment by Ilysen** [comment]
I'm compiling to due some testing on my end, but I decided to go ahead and finish the overhaul of the flags as well and include them in this PR. I'll push a commit once I'm sure it works as intended.

**Comment by Rail-Runner** [comment]
Looking at the flags implemented, wouldn't it be better to go further and make water resistance a JSON value (not a flag) which can be anything from 0 to 100, reflecting the percent of turns on which it protects from rain? Would be more granular this way.

**Comment by Ilysen** [comment]
I think that would be an even-larger overhaul of balance, and I'd like to reserve that for a later pull request if at all possible.

**Comment by Ilysen** [comment]
Going on indefinite break - anyone who wants to pick this up can.

**Pull request closed without merging**